rclcpp/include/rclcpp/node_interfaces/node_base.hpp

```diff
@@ -16,26 +16,52 @@
 #define RCLCPP__NODE_INTERFACES__NODE_BASE_HPP_
 
 #include <memory>
+#include <mutex>
 #include <string>
+#include <unordered_map>
 #include <vector>
 
 #include "rcl/node.h"
+#include "rclcpp/callback_group.hpp"
 #include "rclcpp/context.hpp"
 #include "rclcpp/macros.hpp"
 #include "rclcpp/node_interfaces/node_base_interface.hpp"
 #include "rclcpp/visibility_control.hpp"
 
+
 namespace rclcpp
 {
 namespace node_interfaces
 {
 
+void global_for_each_callback_group(
+  NodeBaseInterface * node_base_interface,
+  const NodeBaseInterface::CallbackGroupFunction & func);
+
+// Class to hold the global map of mutexes
+class map_of_mutexes final
+{
+public:
+  // Methods need to be protected by internal mutex
+  void create_mutex_of_nodebase(const NodeBaseInterface * nodebase);
+  std::shared_ptr<std::mutex>
+  get_mutex_of_nodebase(const NodeBaseInterface * nodebase);
+  void delete_mutex_of_nodebase(const NodeBaseInterface * nodebase);
+
+private:
+  // Members
+  std::unordered_map<const NodeBaseInterface *, std::shared_ptr<std::mutex>> data;
+  std::mutex internal_mutex;
```

Comment:
Nit: all of our member variables end with `_`, so this should be:

```suggestion
  std::unordered_map<const NodeBaseInterface *, std::shared_ptr<std::mutex>> data_;
  std::mutex internal_mutex_;
```

(obviously you'll have to update the code as well)